server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -86,19 +85,13 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             return YES_AUTO_EXPAND_ALL;
         }
         if (node.node() != null) {
-            assert Strings.hasLength(node.node().getHostAddress()) : node;
-            for (RoutingNode checkNode : allocation.routingNodes()) {
-                if (checkNode.node() == null) {
-                    continue;
-                }
-                // check if its on the same host as the one we want to allocate to
-                assert Strings.hasLength(checkNode.node().getHostAddress()) : checkNode;
-                if (checkNode.node().getHostAddress().equals(node.node().getHostAddress())) {
-                    for (ShardRouting assignedShard : assignedShards) {
-                        if (checkNode.nodeId().equals(assignedShard.currentNodeId())) {
-                            return allocation.debugDecision() ? debugNoAlreadyAllocatedToHost(node, allocation) : Decision.NO;
-                        }
-                    }
+            for (ShardRouting assignedShard : assignedShards) {
+                if (allocation.nodes()
+                    .getDataNodes()
```

Comment:
I think `DiscoveryNodes#get(String)` is sufficient, no need for `getDataNodes()` first?

```suggestion
```

Reply:
Good point. I'll address it.